### PR TITLE
Refactor OptionManager to be a short class with a bunch of pure helper functions.

### PR DESCRIPTION
**Pull request by loganfsmyth**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | N
| Major: Breaking Change?  | N
| Minor: New Feature?      | N
| Deprecations?            | 
| Spec Compliancy?         | 
| Tests Added/Pass?        | 
| Fixed Tickets            |
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | 

Last PR before I post the changes to OptionManager for caching. This PR is basically slowly moving things around and simplifying OptionManager, so it no longer has all kinds of hard to follow statics. It's a simple class 100 lines long, with a bunch of pure functions that it uses to slowly build up the config info.

Review-wise it might be easier to just look at the final option-manager.js file to say what you think?

### CODE REVIEW

File: packages/babel-core/src/config/option-manager.js [hunk 1]
```diff
@@ -74,340 +61,398 @@ const optionNames = new Set([
...
-    }
+  const plugin = Object.assign({}, pluginObject, {
+    visitor: clone(pluginObject.visitor || {}),
+  });
+
+  traverse.explode(plugin.visitor);
+
+  let inheritsDescriptor;
+  let inherits;
+  if (plugin.inherits) {
```
Comment (xtuc) [inline packages/babel-core/src/config/option-manager.js]: I had the discussion in a PR last time, could we support multiple inheritance?

```js
{
  inherits: require('foo')
}
```

vs

```js
{
  inherits: [
    require('foo'),
    require('bar'),
  ]
}
```

File: packages/babel-core/src/config/option-manager.js [hunk 2]
```diff
@@ -74,340 +61,398 @@ const optionNames = new Set([
...
-  }
+  return function(...args) {
+    for (const fn of fns) {
+      fn.apply(this, args);
+    }
+  };
 }
 
-OptionManager.memoisedPlugins = [];
+function createBareOptions() {
```
Comment (xtuc) [inline packages/babel-core/src/config/option-manager.js]: This is more `createInitialOptions`?

Comment (loganfsmyth) [inline packages/babel-core/src/config/option-manager.js]: Yeah, I just left the name we had in there already, but I can change it too.

File: packages/babel-core/src/config/option-manager.js (hunk 1 above)
Comment (loganfsmyth) [inline packages/babel-core/src/config/option-manager.js]: Yeah, definitely something we could add.

Comment (loganfsmyth) [inline packages/babel-core/src/config/option-manager.js]: I think I'll leave it for a future PR, but it wouldn't be too bad to add if we want.

Comment (xtuc) [inline packages/babel-core/src/config/option-manager.js]: Yes sure. I will note this in my todo list.

File: packages/babel-core/test/api.js [hunk 3]
```diff
@@ -146,7 +146,7 @@ describe("api", function () {
           plugins: [__dirname + "/../../babel-plugin-syntax-jsx", false],
         });
       },
-      /TypeError: \[BABEL\] unknown: Falsy value found in plugins/
```
Comment (hzoo) [inline packages/babel-core/test/api.js]: So we are ok just saying this is "falsy"?

Comment (loganfsmyth) [inline packages/babel-core/test/api.js]: It does append the value itself now, before it didn't.

Comment (hzoo) [inline packages/babel-core/test/api.js]: nvm I was talking about it saying it was a plugin/preset before but maybe that's not a big deal
